core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/RareTopNOperator.java

```diff
@@ -0,0 +1,257 @@
+
+    private final Map<GroupKey, List<HashMap<FieldKey, Integer>>> groupListMap = new HashMap<>();
+
+    /**
+     * Push the BindingTuple to Group. Two functions will be applied to each BindingTuple to
+     * generate the {@link GroupKey} and {@link FieldKey} GroupKey = GroupKey(bindingTuple),
+     * FieldKey = FieldKey(bindingTuple)
+     */
+    public void push(ExprValue inputValue) {
+      GroupKey groupKey = new GroupKey(inputValue);
+      FieldKey fieldKey = new FieldKey(inputValue);
+      groupListMap.computeIfAbsent(groupKey, k -> {
+        List<HashMap<FieldKey, Integer>> list = new ArrayList<>();
+        HashMap<FieldKey, Integer> map = new HashMap<>();
+        map.put(fieldKey, 1);
+        list.add(map);
+        return list;
+      });
+      groupListMap.computeIfPresent(groupKey, (key, fieldList) -> {
+        fieldList.forEach(map -> {
+          map.computeIfAbsent(fieldKey, f -> 1);
+          map.computeIfPresent(fieldKey, (field, count) -> {
+            return count + 1;
+          });
+        });
+        return fieldList;
+      });
+    }
+
+    /**
+     * Get the list of {@link BindingTuple} for each group.
+     */
+    public List<ExprValue> result() {
+      ImmutableList.Builder<ExprValue> resultBuilder = new ImmutableList.Builder<>();
+
+      groupListMap.forEach((groups, list) -> {
+        Map<String, ExprValue> map = new LinkedHashMap<>();
+        list.forEach(fieldMap -> {
+          List<FieldKey> result = find(fieldMap);
+          result.forEach(field -> {
+            map.putAll(groups.groupKeyMap());
+            map.putAll(field.fieldKeyMap());
+            resultBuilder.add(ExprTupleValue.fromExprValueMap(map));
+          });
+        });
+      });
+
+      return resultBuilder.build();
+    }
+
+    /**
+     * Get a list of result.
+     */
+    public List<FieldKey> find(HashMap<FieldKey, Integer> map) {
+      Comparator<Map.Entry<FieldKey, Integer>> valueComparator;
+      if (rareTopFlag) {
+        valueComparator = Map.Entry.comparingByValue(Comparator.reverseOrder());
+      } else {
+        valueComparator = Map.Entry.comparingByValue();
+      }
+
+      return map.entrySet().stream().sorted(valueComparator).limit(noOfResults)
+          .map(Map.Entry::getKey).collect(Collectors.toList());
+    }
+  }
+
+  /**
+   * Field Key.
+   */
+  @EqualsAndHashCode
+  @VisibleForTesting
+  public class FieldKey {
+
+    private final List<ExprValue> fieldByValueList;
+
+    /**
+     * FieldKey constructor.
+     */
+    public FieldKey(ExprValue value) {
+      this.fieldByValueList = new ArrayList<>();
+      for (Expression fieldExpr : fieldExprList) {
+        this.fieldByValueList.add(fieldExpr.valueOf(value.bindingTuples()));
+      }
+    }
+
+    /**
+     * Return the Map of field and field value.
+     */
+    public Map<String, ExprValue> fieldKeyMap() {
+      Map<String, ExprValue> map = new LinkedHashMap<>();
+      for (int i = 0; i < fieldExprList.size(); i++) {
```

Comment:
Are you using guava? Can you use Streams.zip().collect(Collectors.toMap(...))?

Reply:
Replaced with Streams.zip()...